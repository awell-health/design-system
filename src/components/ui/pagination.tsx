import { Icon } from "./icon";
import { Button } from "./button";
import { useState } from "react";
import { cn } from "@/lib/utils";

export interface PaginationProps {
  totalCount: number;
  onPageChange: (page: number) => void;
  perPage?: number;
  maxButtonsToShow?: number;
}

function Pagination(props: PaginationProps) {
  const { totalCount, onPageChange, perPage = 10, maxButtonsToShow = 5 } = props;
  const [currentPage, setCurrentPage] = useState<number>(0);
  const numberOfPages = Math.ceil(totalCount / perPage);

  const handleClick = (page: number) => {
    setCurrentPage(page);
    onPageChange(page + 1);
  };

  const nextButtonDisabled = currentPage === numberOfPages - 1;
  const prevButtonDisabled = currentPage === 0;
  const showDivider = maxButtonsToShow < numberOfPages;

  const renderItems = () => {
    const buttonsToShow = showDivider ? maxButtonsToShow : numberOfPages;
    const dividerPosition = Math.ceil(buttonsToShow / 2) - 1;
    let items = [...Array(numberOfPages).keys()];

    if (showDivider) {
      const firstItems = Math.round(maxButtonsToShow / 2)
      const lastItems = maxButtonsToShow - firstItems

      items = [
        ...items.slice(0, firstItems),
        ...items.slice(Math.max(items.length - lastItems, 0))
      ]
    }

    return items.map((i) => (
      <>
        <Button
          key={i}
          variant={currentPage === i ? "secondaryBlue" : "ghost"}
          onClick={() => handleClick(i)}
        >
          {i + 1}
        </Button>
        {showDivider && i === dividerPosition && (
          <span key='divider' className="flex h-8 mt-1">...</span>
        )}
      </>
    ));
  };

  return (
    <div
      className={
        "w-full px-6 py-4 bg-white justify-between items-start inline-flex"
      }
    >
      {numberOfPages > 1 && (
        <>
          <div className="w-32">
            <Button
              variant="secondary"
              onClick={() => handleClick(currentPage - 1)}
              disabled={prevButtonDisabled}
            >
              <Icon
                icon="RiArrowLeftLine"
                size={16}
                className={cn(prevButtonDisabled && "fill-slate-200")}
              />{" "}
              Previous
            </Button>
          </div>
          <div className="flex grow gap-2 justify-center">{renderItems()}</div>
          <div className="w-32">
            <Button
              className={cn(nextButtonDisabled && "fill-slate-200")}
              disabled={nextButtonDisabled}
              variant="secondary"
              onClick={() => handleClick(currentPage + 1)}
            >
              Next <Icon icon="RiArrowRightLine" size={16} />
            </Button>
          </div>
        </>
      )}
    </div>
  );
}

export { Pagination };
