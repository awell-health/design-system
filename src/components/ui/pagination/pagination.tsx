import { Fragment, useState } from 'react';
import { cn } from '@/lib/utils';
import { Button } from '../button';
import { Icon } from '../icon';

interface PaginationProps {
  totalCount: number;
  onPageChange: (page: number) => void;
  perPage?: number;
}

function Pagination(props: PaginationProps) {
  const { totalCount, onPageChange, perPage = 10 } = props;
  const [currentPage, setCurrentPage] = useState<number>(0);
  const edgeButtonsToShow = 3;
  const numberOfPages = Math.ceil(totalCount / perPage);

  const handleClick = (page: number) => {
    setCurrentPage(page);
    onPageChange(page + 1);
  };

  const nextButtonDisabled = currentPage === numberOfPages - 1;
  const prevButtonDisabled = currentPage === 0;
  const showDivider = edgeButtonsToShow * 2 < numberOfPages;

  const renderItems = () => {
    let items: Array<number | string> = [...Array(numberOfPages).keys()];

    if (showDivider) {
      const displayCurrentPage =
        currentPage >= edgeButtonsToShow && currentPage < numberOfPages - edgeButtonsToShow;

      const displayPrefixDivider = currentPage > edgeButtonsToShow + 1;

      const displayPrevPage =
        (displayCurrentPage && currentPage > edgeButtonsToShow) ||
        currentPage === items.length - edgeButtonsToShow;

      const displayNextPage =
        (displayCurrentPage && currentPage < items.length - edgeButtonsToShow - 1) ||
        currentPage === edgeButtonsToShow - 1;

      const displaySuffixDivider = currentPage < items.length - edgeButtonsToShow - 2;

      items = [
        ...items.slice(0, edgeButtonsToShow),
        ...(displayPrefixDivider ? ['prefix-divider'] : []),
        ...(displayPrevPage ? [currentPage - 1] : []),
        ...(displayCurrentPage ? [currentPage] : []),
        ...(displayNextPage ? [currentPage + 1] : []),
        ...(displaySuffixDivider ? ['suffix-divider'] : []),
        ...items.slice(Math.max(items.length - edgeButtonsToShow, 0))
      ];
    }

    return items.map((i) => (
      <Fragment key={i}>
        {typeof i === 'string' && i === 'prefix-divider' && (
          <span key={i} className='flex h-8 mt-1'>
            ...
          </span>
        )}
        {typeof i === 'number' && (
          <Button
            key={i}
            variant={currentPage === i ? 'secondaryBlue' : 'ghost'}
            onClick={(e) => {
              e.preventDefault();
              handleClick(i);
            }}
          >
            {i + 1}
          </Button>
        )}
        {typeof i === 'string' && i === 'suffix-divider' && (
          <span key={i} className='flex h-8 mt-1'>
            ...
          </span>
        )}
      </Fragment>
    ));
  };

  return (
    <div className={'w-full px-6 py-4 bg-white justify-between items-start inline-flex'}>
      {numberOfPages > 1 && (
        <>
          <Button
            variant='secondary'
            onClick={(e) => {
              e.preventDefault();
              handleClick(currentPage - 1);
            }}
            disabled={prevButtonDisabled}
          >
            <Icon
              icon='RiArrowLeftLine'
              size={16}
              className={cn(prevButtonDisabled && 'fill-slate-200')}
            />{' '}
            Previous
          </Button>
          <div className='flex grow gap-2 justify-center'>{renderItems()}</div>
          <Button
            className=''
            disabled={nextButtonDisabled}
            variant='secondary'
            onClick={(e) => {
              e.preventDefault();
              handleClick(currentPage + 1);
            }}
          >
            Next{' '}
            <Icon
              icon='RiArrowRightLine'
              size={16}
              className={cn(nextButtonDisabled && 'fill-slate-200')}
            />
          </Button>
        </>
      )}
    </div>
  );
}

export { Pagination };
