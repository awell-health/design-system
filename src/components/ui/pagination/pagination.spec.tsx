import { render } from '@testing-library/react';
import { expect, it, describe, vi } from 'vitest';
import { Pagination } from './pagination';

describe('Pagination', () => {
  const subject = (props = {}) => {
    const defaultProps = {
      totalCount: 86,
      onPageChange: vi.fn()
    };

    return render(<Pagination {...defaultProps} {...props} />);
  };

  it('match snapshot', () => {
    const { container } = subject();
    expect(container).toMatchSnapshot();
  });
});
