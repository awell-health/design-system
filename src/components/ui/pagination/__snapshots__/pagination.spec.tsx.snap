// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Pagination > match snapshot 1`] = `
<div>
  <div
    class="w-full px-6 py-4 bg-white justify-between items-start inline-flex"
  >
    <button
      class="btn btn-secondary bg-white border-slate-300 hover:bg-slate-100 active:bg-slate-100 disabled:text-slate-200 disabled:border-slate-200 disabled:bg-white btn-sm rounded-md px-3.5 py-2"
      disabled=""
    >
      <svg
        class="remixicon fill-slate-200"
        fill="default"
        height="16"
        viewBox="0 0 24 24"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M7.82843 10.9999H20V12.9999H7.82843L13.1924 18.3638L11.7782 19.778L4 11.9999L11.7782 4.22168L13.1924 5.63589L7.82843 10.9999Z"
        />
      </svg>
       
      Previous
    </button>
    <div
      class="flex grow gap-2 justify-center"
    >
      <button
        class="btn text-blue-600 bg-blue-50 border-blue-200 hover:bg-blue-100 hover:border-blue-300 active:bg-blue-100 active:border-blue-300 disabled:bg-blue-50 disabled:border-blue-200 disabled:text-blue-200 btn-sm rounded-md px-3.5 py-2"
      >
        1
      </button>
      <button
        class="btn btn-ghost text-slate-600 hover:bg-slate-100 active:bg-slate-100 disabled:text-slate-200 disabled:bg-transparent btn-sm rounded-md px-3.5 py-2"
      >
        2
      </button>
      <button
        class="btn btn-ghost text-slate-600 hover:bg-slate-100 active:bg-slate-100 disabled:text-slate-200 disabled:bg-transparent btn-sm rounded-md px-3.5 py-2"
      >
        3
      </button>
      <span
        class="flex h-8 mt-1"
      >
        ...
      </span>
      <button
        class="btn btn-ghost text-slate-600 hover:bg-slate-100 active:bg-slate-100 disabled:text-slate-200 disabled:bg-transparent btn-sm rounded-md px-3.5 py-2"
      >
        13
      </button>
      <button
        class="btn btn-ghost text-slate-600 hover:bg-slate-100 active:bg-slate-100 disabled:text-slate-200 disabled:bg-transparent btn-sm rounded-md px-3.5 py-2"
      >
        14
      </button>
      <button
        class="btn btn-ghost text-slate-600 hover:bg-slate-100 active:bg-slate-100 disabled:text-slate-200 disabled:bg-transparent btn-sm rounded-md px-3.5 py-2"
      >
        15
      </button>
    </div>
    <button
      class="btn btn-secondary bg-white border-slate-300 hover:bg-slate-100 active:bg-slate-100 disabled:text-slate-200 disabled:border-slate-200 disabled:bg-white btn-sm rounded-md px-3.5 py-2"
    >
      Next
       
      <svg
        class="remixicon "
        fill="default"
        height="16"
        viewBox="0 0 24 24"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M16.1716 10.9999L10.8076 5.63589L12.2218 4.22168L20 11.9999L12.2218 19.778L10.8076 18.3638L16.1716 12.9999H4V10.9999H16.1716Z"
        />
      </svg>
    </button>
  </div>
</div>
`;
